tests/phpunit/CRM/ManualDirectDebit/Hook/CalculateContributionReceiveDate/OtherContributionTest.php

```diff
@@ -0,0 +1,215 @@
+ * @group headless
+ */
+class CRM_ManualDirectDebit_Hook_CalculateContributionReceiveDate_OtherContributionTest extends BaseHeadlessTest {
+  /**
+   * Default direct debit settings that will be used for tests.
+   *
+   * @var array
+   */
+  private $defaultDDSettings = [
+    'default_reference_prefix' => 'PRE-',
+    'minimum_reference_prefix_length' => 4,
+    'new_instruction_run_dates' => [1],
+    'payment_collection_run_dates' => [5],
+    'minimum_days_to_first_payment' => 1,
+    'second_instalment_date_behaviour' => SettingsManager::SECOND_INSTALMENT_BEHAVIOUR_ONE_MONTH_AFTER,
+  ];
+
+  /**
+   * Default parameters used to create the contribution of a payment plan.
+   *
+   * @var array
+   */
+  private $defaultContributionParams = [
+    'is_pay_later' => TRUE,
+    'skipLineItem' => 1,
+    'skipCleanMoney' => TRUE,
+    'fee_amount' => 0,
+    'payment_instrument_id' => 'direct_debit',
+  ];
+
+  /**
+   * Helper function to create memberships and its default price field value.
+   *
+   * @param array $params
+   *
+   * @return \stdClass
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function createMembershipType($params) {
+    $membershipType = MembershipTypeFabricator::fabricate($params);
+    $priceFieldValue = civicrm_api3('PriceFieldValue', 'get', [
+      'sequential' => 1,
+      'membership_type_id' => $membershipType['id'],
+      'options' => ['limit' => 1],
+    ])['values'][0];
+
+    $result = new stdClass();
+    $result->membershipType = $membershipType;
+    $result->priceFieldValue = $priceFieldValue;
+
+    return $result;
+  }
+
+  /**
+   * Builds a mock class to manage DD settings.
+   *
+   * @return \CRM_ManualDirectDebit_Common_SettingsManager
+   */
+  private function buildSettingsManagerMock($settings) {
+    $settingsManager = $this->createMock(SettingsManager::class);
+    $settingsManager
+      ->method('getManualDirectDebitSettings')
+      ->willReturn(array_merge($this->defaultDDSettings, $settings));
+
+    return $settingsManager;
+  }
+
+  /**
+   * Builds mock receive date calculator object.
+   *
+   * @param string $dayNextMonth
+   *
+   * @return \CRM_MembershipExtras_Service_InstallmentReceiveDateCalculator
+   * @throws \Exception
+   */
+  private function buildReceiveDateCalculatorMock($dayNextMonth) {
+    $calculator = $this->createMock(ReceiveDateCalculator::class);
+    $calculator
+      ->method('getSameDayNextMonth')
+      ->willReturn(new DateTime($dayNextMonth));
+
+    return $calculator;
+  }
+
+  /**
+   * Configures a payment plan to be used on tests.
+   *
+   * @param string $membershipStartDate
+   * @param string $firstInstalmentReceiveDate
+   * @param int $numberOfContributions
+   * @param array $params
+   *
+   * @return mixed
+   * @throws \CiviCRM_API3_Exception
+   */
+  private function setupPlan($membershipStartDate, $firstInstalmentReceiveDate, $numberOfContributions, array $params) {
+    $mainMembershipType = $this->createMembershipType([
```

Comment:
cannot you just use the payment plan fabricator in ME extension?

Reply:
I am using the fabricator in ME extension:

https://github.com/compucorp/uk.co.compucorp.manualdirectdebit/blob/975d8b2d22f4681b3863fbf34e8b7f481574e298/tests/phpunit/CRM/ManualDirectDebit/Hook/CalculateContributionReceiveDate/OtherContributionTest.php#L54

https://github.com/compucorp/uk.co.compucorp.manualdirectdebit/blob/975d8b2d22f4681b3863fbf34e8b7f481574e298/tests/phpunit/CRM/ManualDirectDebit/Hook/CalculateContributionReceiveDate/OtherContributionTest.php#L4

Reply:
no I meant the PaymentPlanOrder fabricator (CRM_MembershipExtras_Test_Fabricator_PaymentPlanOrder) instead of creating all these records manually

Reply:
Oh, got it, I just needed better control over the membership and contribution dates, and especially, I needed to be able to have a plan half set-up, not completeley created, but stop for example at the 4th contribution, so I could run the test to verify the calculation for the 5th contribution would result in the expected date.